packages/ts-json-api/src/types/resourceObjects.ts

```diff
@@ -66,11 +66,11 @@ export interface Relationship<
         | ResourceObject
         | ResourceObject[]
 > {
-    data?: D extends ResourceObject
+    data?: (D extends ResourceObject
         ? ResourceIdentifier<D>
         : D extends ResourceObject[]
         ? ResourceIdentifier<D[0]>[]
-        : ResourceIdentifier | ResourceIdentifier[];
+        : ResourceIdentifier | ResourceIdentifier[]) | null;
```

Comment:
Per the spec, empty to-many relationships should be an empty array. So we should only allow `null` if we know it is an empty to-one relationship, or if we don't know what kind of relationship it is.

```
    data?: D extends ResourceObject
        ? ResourceIdentifier<D> | null
        : D extends ResourceObject[]
        ? ResourceIdentifier<D[0]>[] // <- to-many relationship, null not allowed
        : ResourceIdentifier | ResourceIdentifier[] | null;
```